### PR TITLE
Use unique table locations in Iceberg by default

### PR DESCRIPTION
**Pull request by alexjo2144**
## Description

Enable the unique table location feature by default in the Iceberg connector

> Is this change a fix, improvement, new feature, refactoring, or other?

Enabling an existing feature by default

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Iceberg connector

> How would you describe this change to a non-technical end user or system administrator?

## Related issues, pull requests, and links

Fixes: https://github.com/trinodb/trino/issues/5632

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Use unique table locations in Iceberg by default
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java [hunk 1]
```diff
@@ -46,7 +46,7 @@
     private HiveCompressionCodec compressionCodec = ZSTD;
     private boolean useFileSizeFromMetadata = true;
     private int maxPartitionsPerWriter = 100;
-    private boolean uniqueTableLocation;
+    private boolean uniqueTableLocation = true;
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:49]: The test for the default value should be also adjusted.

**Review by losipiuk — APPROVED** [review]
LGTM % failing test

**Comment by findepi** [comment]
@alexjo2144 you need to rebase

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java [hunk 2]
```diff
@@ -87,6 +84,16 @@ public void testCleanupOnDropTable(Engine tableCreatorEngine, Engine tableDroppe
...
     }
 
+    private String getTableLocation(String tableName)
+    {
+        Pattern locationPattern = Pattern.compile(".*location = 'hdfs://hadoop-master:9000(.*?)'.*", Pattern.DOTALL);
+        Matcher m = locationPattern.matcher((String) onTrino().executeQuery("SHOW CREATE TABLE " + tableName).row(0).get(0));
+        if (m.find()) {
+            return m.group(1);
+        }
+        throw new IllegalArgumentException("Location not found in SHOW CREATE TABLE result");
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java:94]: ISE

**Review by findepi — APPROVED** [review]

**Comment by alexjo2144** [comment]
All set, thank you all for the reviews

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergPartitionEvolution.java [hunk 3]
```diff
@@ -43,19 +44,19 @@ public void testDroppedPartitionField(boolean dropFirst)
 
         onSpark().executeQuery("ALTER TABLE iceberg_test.default.test_dropped_partition_field DROP PARTITION FIELD " + (dropFirst ? "a" : "b"));
 
-        assertThat(onTrino().executeQuery("SHOW CREATE TABLE test_dropped_partition_field"))
-                .containsOnly(
-                        row("CREATE TABLE iceberg.default.test_dropped_partition_field (\n" +
+        Assertions.assertThat((String) onTrino().executeQuery("SHOW CREATE TABLE test_dropped_partition_field").row(0).get(0))
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergPartitionEvolution.java:47]: getOnlyElement

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergPartitionEvolution.java [hunk 4]
```diff
@@ -43,19 +44,19 @@ public void testDroppedPartitionField(boolean dropFirst)
...
+                .matches(
+                        "CREATE TABLE iceberg.default.test_dropped_partition_field \\(\n" +
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergPartitionEvolution.java:49]: use `\\Q`  here too

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java [hunk 5]
```diff
@@ -87,6 +85,17 @@ public void testCleanupOnDropTable(Engine tableCreatorEngine, Engine tableDroppe
         dataFilePaths.forEach(dataFilePath -> assertFileExistence(dataFilePath, false, format("The data file %s removed after dropping the table", dataFilePath)));
     }
 
+    private String getTableLocation(String tableName)
+    {
+        Pattern locationPattern = Pattern.compile(".*location = 'hdfs://hadoop-master:9000(.*?)'.*", Pattern.DOTALL);
+        Matcher m = locationPattern.matcher((String) onTrino().executeQuery("SHOW CREATE TABLE " + tableName).row(0).get(0));
+        if (m.find()) {
+            verify(m.groupCount() == 1, "Location regex only expected one group but found " + m.groupCount());
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java:93]:
```
String location = m.group(1);
verify(!m.find(), "Unexpected second match");
return location;
```

**Comment by alexjo2144** [comment]
Updated. @findepi might be worth running with secrets for Glue?

**Comment by alexjo2144** [comment]
CI: https://github.com/trinodb/trino/issues/12950

**Comment by alexjo2144** [comment]
Rebased for conflicts

**Comment by alexjo2144** [comment]
@losipiuk CI hit https://github.com/trinodb/trino/issues/10631

**Comment by alexjo2144** [comment]
@losipiuk @findepi I think this is good to go whenever you get a chance

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java [hunk 6]
```diff
@@ -443,7 +443,7 @@ public void testSqlFeatures()
                         "WITH (\n" +
                         "   format = 'ORC',\n" +
                         "   format_version = 2,\n" +
-                        "   location = '" + getSchemaDirectory() + "/st_\\E[0-9a-f]+\\Q',\n" +
+                        "   location = '" + getSchemaDirectory() + "/st_\\E[0-9a-f]+\\Q-\\E[0-9a-f]+\\Q',\n" +
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:446]: `\\Q-\\E` -> `-`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergPartitionEvolution.java [hunk 7]
```diff
@@ -43,19 +45,19 @@ public void testDroppedPartitionField(boolean dropFirst)
 
         onSpark().executeQuery("ALTER TABLE iceberg_test.default.test_dropped_partition_field DROP PARTITION FIELD " + (dropFirst ? "a" : "b"));
 
-        assertThat(onTrino().executeQuery("SHOW CREATE TABLE test_dropped_partition_field"))
-                .containsOnly(
-                        row("CREATE TABLE iceberg.default.test_dropped_partition_field (\n" +
+        Assertions.assertThat((String) getOnlyElement(getOnlyElement(onTrino().executeQuery("SHOW CREATE TABLE test_dropped_partition_field").rows())))
```
Comment (losipiuk) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergPartitionEvolution.java:48]: static import? not possible for some reason?

**Review by losipiuk — APPROVED** [review]

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergPartitionEvolution.java (hunk 7 above)
Comment (findepi) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergPartitionEvolution.java:48]: in PT, we use tempto's `assertThat`, so assertj's (rare) ends up without static import

**Review by homar — APPROVED** [review]

**Comment by alexjo2144** [comment]
Checks are passing :heavy_check_mark:

**Pull request merged**